src/Controller/ExportController.php

```diff
@@ -95,9 +103,43 @@ public function export(Request $request, ?SalesChannelContext $context): Respons
             return $errorResponse;
         }
 
+        $excludeProductGroups = filter_var($request->get('excludeProductGroups'), FILTER_VALIDATE_BOOLEAN);
+        $dynamicProductGroupService = $this->getDynamicProductGroupService();
+        if (!$excludeProductGroups && !$dynamicProductGroupService->isDynamicProductGroupWarmedUp()) {
+            $errorMessage = 'Dynamic Product Groups have not been warmed up yet. ';
+            $errorMessage .= 'This may cause missing categories! ';
+            $errorMessage .= "Warm them up by calling the route '/findologic/dynamic-product-groups', ";
+            $errorMessage .= 'or disable fetching of Dynamic Product Groups by adding the query parameter ';
+            $errorMessage .= "'excludeProductGroups=true'";
+
+            return new JsonResponse(['error' => $errorMessage], Response::HTTP_PRECONDITION_REQUIRED);
```

Comment:
I would suggest we create a separate class for this response, as the contents here bloat the actual code and distracts from the real implementation. E.g.

```php
class PreconditionFailedResponse extends JsonResponse
{
    public const CODE_DYNAMIC_PRODUCT_GROUPS_NOT_WARMED_UP = 0;

    public function __construct(int $errorCode, /* ... */)
    {
        switch ($errorCode) {
             case self::CODE_DYNAMIC_PRODUCT_GROUPS_NOT_WARMED_UP:
                 $errorMessage = '...';
                 break;
             default:
                 $errorMessage = 'Unknown';
        }
        
        parent::__construct(/* ... */);
    }
}
```

I suggest adding it inside the `Export` folder and create a new folder called `Responses`

Reply:
As this is also considered an "error response", please make sure that `$this->validate()` is responsible for it.